lepton/azure_network.go

```diff
@@ -69,20 +89,157 @@ func (a *Azure) CreateNIC(ctx context.Context, location string, vnetName, subnet
 	}
 
 	fmt.Printf("\nCreated NIC for instance")
-	return future.Result(nicClient)
+	nic, err = future.Result(*nicClient)
+	return
+}
+
+// DeleteNIC deletes network interface controller
+func (a *Azure) DeleteNIC(ctx *Context, nic *network.Interface) error {
+	logger := ctx.logger
+
+	nicClient, err := a.getNicClient()
+	if err != nil {
+		return err
+	}
+
+	logger.Info("Deleting %s...", *nic.ID)
+	nicName := getAzureResourceNameFromID(*nic.ID)
+	nicDeleteTask, err := nicClient.Delete(context.TODO(), a.groupName, nicName)
+	if err != nil {
+		logger.Error("error deleting network interface controller: %v", err)
+		return errors.New("error deleting network interface controller")
+	}
+
+	err = nicDeleteTask.WaitForCompletionRef(context.TODO(), nicClient.Client)
+	if err != nil {
+		logger.Error("error waiting for network interface controller deleting: %v", err)
+		return errors.New("error waiting for network interface controller deleting")
+	}
+
+	return nil
+}
+
+// DeletePublicIPs deletes array of ip configurations
+func (a *Azure) DeletePublicIPs(ctx *Context, ips *[]network.InterfaceIPConfiguration) error {
+	logger := ctx.logger
+
+	ipClient, err := a.getIPClient()
+	if err != nil {
+		return err
+	}
+
+	for _, ip := range *ips {
+		if ip.PublicIPAddress.ID != nil {
+			ipID := getAzureResourceNameFromID(*ip.PublicIPAddress.ID)
+
+			logger.Info("Deleting %s...", *ip.PublicIPAddress.ID)
+			deleteIPTask, err := ipClient.Delete(context.TODO(), a.groupName, ipID)
+			if err != nil {
+				logger.Error("error deleting ip: %v", err)
+				return errors.New("error deleting ip")
+			}
+
+			err = deleteIPTask.WaitForCompletionRef(context.TODO(), ipClient.Client)
+			if err != nil {
+				logger.Error("error waiting for ip deletion: %v", err)
+				return errors.New("error waiting for ip deletion")
+			}
+		}
+	}
+
+	return nil
 }
 
-func (a *Azure) getIPClient() network.PublicIPAddressesClient {
+// DeleteNetworkSecurityGroup deletes virtual networks and subnets associated with the security group and the security group itself
+func (a *Azure) DeleteNetworkSecurityGroup(ctx *Context, securityGroupID string) error {
+	logger := ctx.logger
+
+	nsgClient, err := a.getNsgClient()
+	if err != nil {
+		return err
+	}
+	subnetsClient, err := a.getSubnetsClient()
+	if err != nil {
+		return err
+	}
+	vnetClient, err := a.getVnetClient()
+	if err != nil {
```

Comment:
This is an example of boilerplate (error handling) and unnecessary code executed referred in a previous comment.

`getNsgClient`, `getSubnetsClient` and `getVnetClient` requires a resource management authorizer. Each one of these methods is doing a network call to get it.

Reply:
https://github.com/nanovms/ops/issues/778